scipy/optimize/_differentiable_functions.py

```diff
@@ -145,18 +148,23 @@ def update_hess():
 
         if isinstance(hess, HessianUpdateStrategy):
             def update_x(x):
+                x = np.atleast_1d(x).astype(float)
+                if np.any((x < self.lb) | (x > self.ub)):
+                    raise ValueError("`updated x` violates bound constraints.")
```

Comment:
I think I would avoid raising an error here. We don't want the algorithm to stop when a point out of the bounds is evaluated.

I know you are catching this error on ``tr_interior_point.py`` but this class should be general and we might want to use it for other algorithms. 

maybe just avoiding updating the gradient and the hessian, is not a best solution? We can possibly set evaluated function to `inf`?

Reply:
Perhaps we can avoid this if we add some logic to scalarfunction and vectorfunction objects to not attempt any hessian (and perhaps gradient and jacobian) updates unless we confirm x is in the feasible region. I'll look into this approach.

Reply:
My initial thought is that if we're going to add bounds to any of the `_differentiable_functions` it should raise a ValueError if a position outside the bounds is requested.
However, that would be a last resort, I'd much prefer that the optimizer algorithm doesn't call into this code with an out of bounds position in the first place. Can `tr_interior_point` not examine to see if `x` is valid before it requests that a Hessian be updated?

 As @antonior92 these functions are used by several other pieces of code, and they need to remain general.